google-cloud-spanner/src/main/java/com/google/cloud/spanner/Mutation.java

```diff
@@ -402,20 +404,41 @@ private boolean isFloat32NaN(Value value) {
     return value.getType().equals(Type.float32()) && Float.isNaN(value.getFloat32());
   }
 
-  static void toProto(Iterable<Mutation> mutations, List<com.google.spanner.v1.Mutation> out) {
+  // Converts Mutation and returns a random Mutation from the available mutation list based on the
+  // following heuristics:
+  // 1. Prefer mutations other than INSERT, since INSERT mutations may contain autogenerated columns
+  //    whose information is unavailable on the client.
+  // 2. If the list only contains INSERT mutations, select the one with the highest number of
+  // values.
+  static com.google.spanner.v1.Mutation toProtoGetRandomMutation(
+      Iterable<Mutation> mutations, List<com.google.spanner.v1.Mutation> out) {
     Mutation last = null;
     // The mutation currently being built.
     com.google.spanner.v1.Mutation.Builder proto = null;
     // The "write" (!= DELETE) or "keySet" (==DELETE) for the last mutation encoded, for coalescing.
     com.google.spanner.v1.Mutation.Write.Builder write = null;
     com.google.spanner.v1.KeySet.Builder keySet = null;
+
+    // Store all the mutations excluding INSERT.
+    List<com.google.spanner.v1.Mutation> allMutationsExcludingInsert = new ArrayList<>();
+    // Stores INSERT mutation with large number of values.
+    com.google.spanner.v1.Mutation largeInsertMutation =
```

Comment:
```suggestion
    com.google.spanner.v1.Mutation largestInsertMutation =
```

Reply:
One very peculiar, but not impossible, case is that the list of mutations only contains empty insert mutations. This could for example be possible for a table that has a default value for the primary key (or allows null for the primary key), and also either has a default value for all other columns, or allows them to be null. Such a case would lead to this method returning `Mutation.getDefaultInstance()` as the largest insert mutation.

The heuristics should therefore also be changed to prefer an insert mutation with zero columns and a table name, over an insert mutation with zero columns and no table name. (Any other non-empty property of a real mutation can also be used instead of the table name for this check).

Can we also add a test for this specific scenario? It is typically something that could also backslide in the future.

Reply:
Thanks Knut for pointing out to this edge case.

I have updated the code to handle this case by adding below lines, and added a test for empty insert mutation scenario.
```
    // If largestInsertMutation is a default instance of Mutation, replace it with the current
    // INSERT mutation, even if it contains zero values.
    if (!largestInsertMutation.hasInsert()) {
      return true;
    }
```

Talking out loud of all the edge cases,
1. If mutations are empty, then this logic never gets called.
2. Table name in a mutation can not be empty.
3. If there are mutations other than INSERT operation, then once of them is chosen at random.
4. If only INSERT mutations, then one with largest value is chosen
5. If only INSERT mutation, with no columns/values set, then this mutation will be chosen as random mutation

Please let me know if there is anything missing.